source/Ship.cpp

```diff
@@ -3760,6 +3760,230 @@ void Ship::AddOutfit(const Outfit *outfit, int count)
+		}
+		else
+		{
+			baseAttributes.Set(targetAttribute.c_str(), newBaseValue);
+			attributes.Set(targetAttribute.c_str(), newValue);
+		}
+
+		// Ensuring the current hull value is changed as well.
+		if(targetAttribute == "hull")
+		{
+			// Adds the modifyAmount and the limiter to the current hull too.
+			hull += modifyAmount + limiter;
+		}
+
+		// Ensuring the current shields value is changed as well.
+		if(targetAttribute == "shields")
+		{
+			// Adds the modifyAmount and the limiter to the current shields too.
+			shields += modifyAmount + limiter;
+		}
+
+		if(targetAttribute == "cargo space")
+		{
+			cargo.SetSize(attributes.Get("cargo space"));
+
+			// Only the player's ships make use of attraction and deterrence.
+			if(isYours)
+				attraction = CalculateAttraction();
+		}
+
+		// If the added or removed attribute is hyperdrive, scram drive, or jump drive capability, then
+		// recalibrate the navigation.
+		if((targetAttribute == "hyperdrive" || targetAttribute == "scram drive" || targetAttribute == "jump drive"))
+			navigation.Calibrate(*this);
+		// Navigation may still need to be recalibrated depending on the drives a ship has.
+		// Only do this for player ships as to display correct information on the map.
+		// Non-player ships will recalibrate before they jump.
+		else if(isYours)
+			navigation.Recalibrate(*this);
+	}
+}
+
+
+
+// Sets an attribute to a specific value. Certain values (mass, drag, space, etc.) have limits.
+void Ship::SetAttribute(string targetAttribute, double setAmount)
+{
+	if(!targetAttribute.empty())
+	{
+		double limiter = 0.;
+		double minAttributeValue = 0.;
+		double originalBaseValue = baseAttributes.Get(targetAttribute);
+		double originalValue = attributes.Get(targetAttribute);
+		// This is to account for potential differences between the value in baseAttributes & Attributes.
+		double originalDifference = originalValue - originalBaseValue;
+		double newBaseValue = setAmount;
+		double newValue = setAmount + originalDifference;
+
+		// double intermediaryValue = originalBaseValue - originalValue;
+
+		// Safety checks to ensure the new value is within parameters.
+		if(newBaseValue < 1. && targetAttribute == "hull")
+		{
+			// This means the minimum value for this attribute is 1.0
+			minAttributeValue = 1.;
+			limiter = minAttributeValue - newBaseValue;
+		}
+		else if(newValue < 1. && targetAttribute == "hull")
+		{
+			// This means the minimum value for this attribute is 1.0
+			minAttributeValue = 1.;
+			limiter = minAttributeValue - newValue;
+		}
+		else if(newBaseValue < 0.01 && targetAttribute == "drag")
+		{
+			// This means the minimum value for these attributes is 0.01
+			minAttributeValue = 0.01;
+			limiter = minAttributeValue - newBaseValue;
+		}
+		else if(newValue < 0.01 && targetAttribute == "drag")
+		{
+			// This means the minimum value for these attributes is 0.01
+			minAttributeValue = 0.01;
+			limiter = minAttributeValue - newValue;
```

Comment:
Just a note I didnt take a good look at thus yet but seems a bit weird to me

Reply:
The idea is that these attributes should never go below 0.01. I don't want to just cap it at 0.01 because it's always possible that attributes and baseattributes are different due to whatever is going on in game. So instead of just doing "if newValue < 0.01, then use 0.01 instead", I'm checking the difference between the two to give the limiter value. Then I can use that as the new modifier to both attributes and base attributes, only decreasing the value as much as it can (and dropping both equally) instead of having them change un-equally.